suzieq/poller/controller/source/nautobot.py

```diff
@@ -0,0 +1,228 @@
+            else:
+                if server.protocol == "http" and ssl_verify:
+                    raise ValueError("Cannot use ssl_verify=True with http host")
+            values["ssl_verify"] = ssl_verify
+            return server
+        elif isinstance(url, NautobotServerModel):
+            return url
+        else:
+            raise ValueError("Unknown input type")
+
+    @validator("token")
+    def validate_token(cls, token):
+        """checks if the token can be load as sensible data"""
+        try:
+            if token == "ask":
+                return token
+            return get_sensitive_data(token)
+        except SensitiveLoadError as e:
+            raise ValueError(e)
+
+
+class Nautobot(Source, InventoryAsyncPlugin):
+    """This class is used to dynamically retrieve the inventory from Nautobot"""
+
+    def __init__(self, config_data: dict, validate: bool = True) -> None:
+        self._status = "init"
+        self._session: api = None
+        self._server: NautobotServerModel = None
+        super().__init__(config_data, validate)
+
+    @classmethod
+    def get_data_model(cls):
+        return NautobotSourceModel
+
+    def _load(self, input_data):
+        # load the server class from the dictionary
+        if not self._validate:
+            input_data["server"] = NautobotServerModel.construct(
+                **input_data.pop("url", {})
+            )
+            input_data["ssl_verify"] = input_data.pop("ssl-verify", False)
+        super()._load(input_data)
+        if self._data.token == "ask":
+            self._data.token = get_sensitive_data(
+                "ask", f"{self.name} Insert Nautobot API token: "
+            )
+        self._server = self._data.server
+        if not self._auth:
+            raise InventorySourceError(
+                f"{self.name} Nautobot must have an "
+                "'auth' set in the 'namespaces' section"
+            )
+
+    def _init_session(self):
+        """Initialize the session property
+
+        Args:
+            headers ([dict]): headers to initialize the session
+        """
+        url_address = (
+            f"{self._server.protocol}://{self._server.host}:" f"{self._server.port}"
+        )
+
+        if not self._session:
+            ssl_option = self._data.ssl_verify
+            self._session = api(
+                url=url_address, token=self._data.token, verify=ssl_option
+            )
+
+    async def get_inventory_list(self) -> List:
+        """Contact Nautobot to retrieve the inventory.
+
+        Raises:
+            InventorySourceError: Unable to connect to the REST server
+
+        Returns:
+            List: inventory list
+        """
+
+        if not self._session:
+            self._init_session()
+        try:
+            if self._data.device_filters:
+                devices = self._session.dcim.devices.filter(**self._data.device_filters)
+            else:
+                devices = self._session.dcim.devices.all()
```

Comment:
In netbox, this function handles NSOT servers behind a proxy. Do you know if the `pynautobot` automatically handles these kind of problems?

Reply:
I'm not sure about that.